### PR TITLE
Mark `experimental` encoding as `arrow.v2`

### PR DESCRIPTION
**Pull request by dominiklohmann**
This makes the previously experimental encoding the new default.

Outstanding tasks:

- [x] Make the experimental encoding the fault encoding
- [x] Remove the old Arrow table slice builder
- [x] Rename the new Arrow table slice builder
- [x] Add utilities for working with nested struct arrays in record batches
- [x] ~Update transform step: `identity`~ No changes required
- [x] Update transform step: `hash`
- [x] ~Update transform step: `count`~ No changes required
- [x] Update transform step: `delete`
- [x] ~Update transform step: `rename`~ No changes required
- [x] ~Update transform step: `select`~ No changes required
- [x] Update transform step: `project`
- [x] Update transform step: `replace`
- [x] Update transform step: `aggregate`

### :memo: Checklist

- [x] All user-facing changes have changelog entries.
- [x] The changes are reflected on [docs.tenzir.com/vast](https://docs.tenzir.com/vast), if necessary.
- [x] The PR description contains instructions for the reviewer, if necessary.

### :dart: Review Instructions

Commit-by-commit.

### CODE REVIEW

**Comment by tobim** [comment]
I was able to fix the linker error with:
```
template auto values(const type& type, const arrow::Array& array) noexcept
  -> detail::generator<data_view>;
```
But clangd still complains, so I guess this won't work with clang.

File: libvast/src/arrow_table_slice.cpp [hunk 1]
```diff
@@ -1128,6 +1128,244 @@ arrow_table_slice<FlatBuffer>::record_batch() const noexcept {
...
+          .arrays = caf::get<type_to_arrow_array_t<record_type>>(
+                      *layer.arrays[index.back()])
+                      .fields(),
+        };
+        nested_layer.fields.reserve(nested_layer.arrays.size());
+        for (auto&& [name, type] :
+             caf::get<record_type>(layer.fields[index.back()].type).fields())
+          nested_layer.fields.push_back({std::string{name}, type});
+        auto nested_index = index;
+        nested_index.push_back(0);
```
Comment (tobim) [inline libvast/src/arrow_table_slice.cpp:1210]: Wdyt of renaming this `index` to `cursor`? That would be a little more precise imo.
`auto{index}.push_back(0)` would have been handy here as well.

File: libvast/src/arrow_table_slice.cpp [hunk 2]
```diff
@@ -1128,6 +1128,244 @@ arrow_table_slice<FlatBuffer>::record_batch() const noexcept {
...
+        VAST_ASSERT(current != sentinel);
+        for (auto&& [field, array] : std::invoke(
+               std::move(current->fun), std::move(layer.fields[index.back()]),
+               std::move(layer.arrays[index.back()]))) {
+          result.fields.push_back(std::move(field));
+          result.arrays.push_back(std::move(array));
+        }
+        ++current;
+      } else if (is_prefix_match) {
+        auto nested_layer = unpacked_layer{
```
Comment (tobim) [inline libvast/src/arrow_table_slice.cpp:1199]: How about an assert that we have a record_type?

File: libvast/test/arrow_table_slice.cpp [hunk 3]
```diff
@@ -654,7 +764,6 @@ TEST(arrow record type to schema roundtrip tp) {
 }
 
 namespace {
-
```
Comment (tobim) [inline libvast/test/arrow_table_slice.cpp]: Accidental delete?

**Review by tobim — COMMENTED** [review]
First review, I only looked at `project` from the utilities patch so far.

File: libvast/src/arrow_table_slice.cpp (hunk 2 above)
Comment (dominiklohmann) [reply libvast/src/arrow_table_slice.cpp:1199]: The `caf::get<record_type>(...)` internally asserts that already:

```cpp
template <vast::complex_type T, int Index>
static const T& caf::sum_type_access<vast::type>::get(
    const vast::type& x, sum_type_token<T, Index> token) {
  VAST_ASSERT(is(x, token));
  return static_cast<const T&>(
   static_cast<const vast::stateful_type_base&>(x));
}
```

**Comment by tobim** [comment]
> > I was able to fix the linker error with:
> > ```c++
> > template auto values(const type& type, const arrow::Array& array) noexcept
> >   -> detail::generator<data_view>;
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > But clangd still complains, so I guess this won't work with clang.
> 
> That makes no sense to me as that is definitely not a function template. It's just a regular old function.

Oh but it is a template, you use `const std::same_as<arrow::Array> auto& array` for the second parameter which is an abbreviated function template with a constraint.

File: libvast/include/vast/arrow_table_slice.hpp [hunk 4]
```diff
@@ -134,6 +134,41 @@ class arrow_table_slice final {
...
+          const std::vector<indexed_transformation>& transformations) noexcept;
+
+/// Removed all unspecified columns from both a VAST layout and an Arrow record
+/// batch.
+/// @pre VAST layout and Arrow schema must match.
+/// @pre Indices must be sorted.
+/// @pre Indices must not be a subset of the following index.
+std::pair<type, std::shared_ptr<arrow::RecordBatch>>
+project(type layout, const std::shared_ptr<arrow::RecordBatch>& batch,
+        const std::vector<offset>& indices) noexcept;
```
[the remark this answers is not in the document]
Comment (dominiklohmann) [reply libvast/include/vast/arrow_table_slice.hpp:179]: It's the analog to Arrow's `arrow::RecordBatch::SelectColumns`, which only takes flat indices.

I would be open to renaming the functions to `select_columns` and `transform_columns`. Does that make more sense to you?

**Comment by dominiklohmann** [comment]
> I'll try to come up with a way to fix this later, I wanna get the aggregate plugin converted first.

I fixed this accidentally when moving everything to the header.

File: plugins/aggregate/aggregate.cpp [hunk 5]
```diff
@@ -26,6 +30,64 @@
...
+
+std::shared_ptr<arrow::RecordBatch>
+unflatten_batch(const vast::type& layout,
+                const arrow::RecordBatch& batch) noexcept {
+  auto columns = arrow::ArrayVector{};
+  columns.reserve(caf::get<record_type>(layout).num_fields());
+  const auto& flattened_columns = batch.columns();
+  auto current_flattened_column = flattened_columns.begin();
+  auto iteration = 0;
+  fmt::print(stderr, "layout: {}\n", caf::get<record_type>(layout));
```
Comment (dominiklohmann) [inline plugins/aggregate/aggregate.cpp:60]: Remove debug print

File: libvast/include/vast/arrow_table_slice_builder.hpp [hunk 6]
```diff
@@ -3,7 +3,7 @@
 //   | |/ / __ |_\ \  / /          Across
 //   |___/_/ |_/___/ /_/       Space and Time
 //
-// SPDX-FileCopyrightText: (c) 2019 The VAST Contributors
+// SPDX-FileCopyrightText: (c) 2021 The VAST Contributors
```
Comment (lava) [inline libvast/include/vast/arrow_table_slice_builder.hpp:6]: The advice from [the internet](https://matija.suklje.name/how-and-why-to-properly-write-copyright-statements-in-your-code) is to not update the date when touching a file.

> the year of the first publication – a good date would be the year in which you created the file and then [do not touch that date anymore](https://matija.suklje.name/how-and-why-to-properly-write-copyright-statements-in-your-code#why-not-bump-the-year-on-change);

File: changelog/unreleased/changes/2159--arrow-v2.md [hunk 7]
```diff
@@ -0,0 +1,11 @@
...
+completely when using the `arrow` encoding, and maps the pattern, address,
+subnet, and enumeration types onto Arrow extension types rather than using the
+underlying representation directly. This change enables the use of the `export
+arrow` command without needing information about VAST' type system.
+
+Transform steps that add or modify columns now add or modify the columns
+in-place rather than at the end, preserving the nesting structure of the
+original data.
+
+The deprecated `msgpack` encoding no longer exists.
```
Comment (lava) [inline changelog/unreleased/changes/2159--arrow-v2.md:11]: Maybe we should clarify that it no longer exists *when importing new data*, but existing msgpack data can still be read and exported.

**Review by lava — APPROVED** [review]
Looks great! We've mostly reviewed this in a video call, and only the changes that @tobim hadn't reviewed yet, so there's just some minor nits below.

File: libvast/include/vast/arrow_table_slice_builder.hpp (hunk 6 above)
Comment (dominiklohmann) [reply libvast/include/vast/arrow_table_slice_builder.hpp:6]: I think this was changed when I renamed it from experimental, so I probably should change it back manually.

**Comment by dominiklohmann** [comment]
@tobim this is ready except for one linker error in the Nix build I do not understand. Can you take a look and push a fix directly? Other than that this is all ready with auto-merge enabled.

```
/nix/store/j6dpgmc0qzsilpic61bvxmfxj31c5igv-x86_64-unknown-linux-musl-binutils-2.35.2/bin/x86_64-unknown-linux-musl-ld: /build/ccU6jcvG.ltrans9.ltrans.o: in function `vast::hash_step::add(vast::type, std::shared_ptr<arrow::RecordBatch>)::{lambda(vast::record_type::field, std::shared_ptr<arrow::Array>)#1}::operator()(vast::record_type::field, std::shared_ptr<arrow::Array>) const':
[796](https://github.com/tenzir/vast/runs/5695270172?check_suite_focus=true#step:5:796)
       > <artificial>:(.text+0xd994): undefined reference to `vast::detail::generator<caf::variant<caf::none_t, bool, vast::integer, unsigned long, double, std::chrono::duration<long, std::ratio<1l, 1000000000l> >, std::chrono::time_point<std::chrono::_V2::system_clock, std::chrono::duration<long, std::ratio<1l, 1000000000l> > >, std::basic_string_view<char, std::char_traits<char> >, vast::pattern_view, vast::address, vast::subnet, unsigned char, vast::container_view_handle<vast::list_view_ptr>, vast::container_view_handle<vast::map_view_ptr>, vast::container_view_handle<vast::record_view_ptr> > > vast::values<arrow::Array>(vast::type const&, arrow::Array const&)'
[797](https://github.com/tenzir/vast/runs/5695270172?check_suite_focus=true#step:5:797)
       > /nix/store/j6dpgmc0qzsilpic61bvxmfxj31c5igv-x86_64-unknown-linux-musl-binutils-2.35.2/bin/x86_64-unknown-linux-musl-ld: <artificial>:(.text+0xdbb1): undefined reference to `vast::detail::generator<caf::variant<caf::none_t, bool, vast::integer, unsigned long, double, std::chrono::duration<long, std::ratio<1l, 1000000000l> >, std::chrono::time_point<std::chrono::_V2::system_clock, std::chrono::duration<long, std::ratio<1l, 1000000000l> > >, std::basic_string_view<char, std::char_traits<char> >, vast::pattern_view, vast::address, vast::subnet, unsigned char, vast::container_view_handle<vast::list_view_ptr>, vast::container_view_handle<vast::map_view_ptr>, vast::container_view_handle<vast::record_view_ptr> > > vast::values<arrow::Array>(vast::type const&, arrow::Array const&)'
```

**Comment by dispanser** [comment]
There's a failing test on my system, related to converting arrow batches from v0/v1 -> v2.

Looking into this right now, I disabled auto-merge. I'm also not so sure about the CI failures, I don't see that on my end.

**Comment by dispanser** [comment]
closing and reopening,, as the labels check is not working

**Pull request merged**